packages/graph-editor/src/hooks/useCopyPaste.ts

```diff
@@ -1,14 +1,129 @@
+import {
+  Node as GraphNode,
+  NodeFactory,
+  SerializedEdge,
+} from '@tokens-studio/graph-engine';
 import { graphEditorSelector } from '@/redux/selectors/graph.js';
-import { useCallback } from 'react';
+import { nodeTypesSelector } from '@/redux/selectors/registry.js';
+import { useCallback, useRef } from 'react';
 import { useReactFlow } from 'reactflow';
+import { useSelectAddedNodes } from '@/hooks/useSelectAddedNodes.js';
 import { useSelector } from 'react-redux';
 import { useToast } from '@/hooks/useToast.js';
 import { v4 as uuidv4 } from 'uuid';
+import { xpos, ypos } from '@/annotations/index.js';
+import pluralizeValue from '@/utils/pluralizeValue.js';
+
+type EdgesByTargetMap = Record<string, Array<SerializedEdge>>;
 
 function useCopyPaste() {
   const graphRef = useSelector(graphEditorSelector);
+  const fullNodeLookup = useSelector(nodeTypesSelector) as Record<
+    string,
+    NodeFactory
+  >;
   const reactFlowInstance = useReactFlow();
   const toast = useToast();
+  const selectAddedNodes = useSelectAddedNodes();
+  const pasteInProgressRef = useRef(false);
+
+  const calculateBoundingBox = useCallback((nodes) => {
+    return nodes.reduce(
+      (acc, node) => ({
+        minX: Math.min(acc.minX, node.annotations?.[xpos] || 0),
+        minY: Math.min(acc.minY, node.annotations?.[ypos] || 0),
+        maxX: Math.max(
+          acc.maxX,
+          (node.annotations?.[xpos] || 0) +
+            (node.annotations?.['ui.width'] || 200),
+        ),
+        maxY: Math.max(
+          acc.maxY,
+          (node.annotations?.[ypos] || 0) +
+            (node.annotations?.['ui.height'] || 100),
+        ),
+      }),
+      {
+        minX: Infinity,
+        minY: Infinity,
+        maxX: -Infinity,
+        maxY: -Infinity,
+      },
+    );
+  }, []);
+
+  const parseClipboardData = useCallback(async () => {
+    try {
+      const text = await navigator.clipboard.readText();
+      const { nodes: sourceNodes, edges: sourceEdges } = JSON.parse(text);
+      return { sourceNodes, sourceEdges };
+    } catch (error) {
+      console.error('Error parsing clipboard data:', error);
+      throw new Error('Invalid clipboard data');
+    }
+  }, []);
+
+  const updateNodeIdsRecursively = useCallback((node, idMapping) => {
+    const newId = uuidv4();
+    idMapping.set(node.id, newId);
+
+    const updatedNode = {
+      ...node,
+      id: newId,
+    };
+
+    // process inner graph if it exists
+    if (node.innergraph && Array.isArray(node.innergraph.nodes)) {
```

Comment:
I wonder about this code. A new node that we have coming is the external subgraph. It's main change over the nested subgraph is that we only store the reference to the graph without serializing it, and retrieve it once during deserialization.

In this case I wonder if we need to prevent going inside the innergraph of that node to prevent changing anything inside it 

Secondly, this is relying on hardcoded knowledge that nodes that contain subgrpahs will expose them specifically with the `.innergraph` property. If we had something that had 2 innergraphs for example, how would we handle that here? It would be very brittle